FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1323 @@
+        auto* TopValue = GetTopWithCache_Slow();
+        Ref RotAmount = IREmit->_Sub(OpSize::i32Bit, GetConstant(8), TopValue);
+        Ref AbridgedFTW = IREmit->_LoadContext(1, GPRClass, offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+        Ref NewAbridgedFTW = IREmit->_Or(OpSize::i32Bit, AbridgedFTW, RotateRight8(Mask, RotAmount));
+        IREmit->_StoreContext(1, GPRClass, NewAbridgedFTW, offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+      }
+    }
+  }
+  { // Set invalid tags
+    uint8_t Mask = StackData.getInvalidMask();
+    if (Mask == 0xff) {
+      IREmit->_StoreContext(1, GPRClass, GetConstant(0), offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+    } else if (Mask != 0) {
+      if (std::popcount(Mask)) {
+        uint8_t BitIdx = __builtin_ctz(Mask);
+        SetX87ValidTag(GetOffsetTopWithCache_Slow(BitIdx), false);
+      } else {
+        // Same rotate right as above but this time on the invalid mask
+        auto* TopValue = GetTopWithCache_Slow();
+        Ref RotAmount = IREmit->_Sub(OpSize::i32Bit, GetConstant(8), TopValue);
+        Ref AbridgedFTW = IREmit->_LoadContext(1, GPRClass, offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+        Ref NewAbridgedFTW = IREmit->_Andn(OpSize::i32Bit, AbridgedFTW, RotateRight8(Mask, RotAmount));
+        IREmit->_StoreContext(1, GPRClass, NewAbridgedFTW, offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+      }
+    }
+  }
+  return TopValue;
+}
+
+std::tuple<Ref, Ref> X87StackOptimization::SplitF64SigExp(Ref Node) {
+  Ref Gpr = IREmit->_VExtractToGPR(8, 8, Node, 0);
+
+  Ref Exp = IREmit->_And(OpSize::i64Bit, Gpr, GetConstant(0x7ff0000000000000LL));
+  Exp = IREmit->_Lshr(OpSize::i64Bit, Exp, GetConstant(52));
+  Exp = IREmit->_Sub(OpSize::i64Bit, Exp, GetConstant(1023));
+  Exp = IREmit->_Float_FromGPR_S(8, 8, Exp);
+  Ref Sig = IREmit->_And(OpSize::i64Bit, Gpr, GetConstant(0x800fffffffffffffLL));
+  Sig = IREmit->_Or(OpSize::i64Bit, Sig, GetConstant(0x3ff0000000000000LL));
+  Sig = IREmit->_VCastFromGPR(8, 8, Sig);
+
+  return std::tuple {Exp, Sig};
+}
+
+void X87StackOptimization::Run(IREmitter* Emit) {
+  FEXCORE_PROFILE_SCOPED("PassManager::x87StackOpt");
+
+  auto CurrentIR = Emit->ViewIR();
+  auto* HeaderOp = CurrentIR.GetHeader();
+  LOGMAN_THROW_AA_FMT(HeaderOp->Header.Op == OP_IRHEADER, "First op wasn't IRHeader");
+
+  if (!HeaderOp->HasX87) {
+    // If there is no x87 in this, just early exit.
+    return;
+  }
+
+  // Initialize IREmit member
+  IREmit = Emit;
+
+  // Run optimization proper
+  for (auto [BlockNode, BlockHeader] : CurrentIR.GetBlocks()) {
+    auto BlockIROp = BlockHeader->CW<FEXCore::IR::IROp_CodeBlock>();
+    // Each time we deal with a new block we need to start over.
+    // The optimization should run per-block
+    Reset();
+
+    for (auto [CodeNode, IROp] : CurrentIR.GetCode(BlockNode)) {
+      IREmit->SetWriteCursor(CodeNode);
+
+      switch (IROp->Op) {
+
+        // Binary ops on the stack
```

Comment:
I think now that they're going through named helpers (HandleBinopStack) we can drop this comment and the blank line above it

Reply:
Yes to these and all the others like it. :)